### PR TITLE
Rust daemon http transport

### PR DESCRIPTION
**Pull request by benjamreis**

### CODE REVIEW

File: rust/cli/README.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+From the `rust` folder:
+```
+cargo build -p xenops
+```
+
+# Usage
+
+The binary is located in: `xenops-ng/rust/target/debug`.
+- `xenops-cli {{pause|unpause|shutdown}} <integer>`: pause/unpause/ shutdown a domain, the integer arg must be a valid domain id.
```
Comment (Wescoeur) [inline rust/cli/README.md]: `Pause/unpause/shutdown`

File: rust/cli/README.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+- `xenops-cli domain-list`: list all domains ids.
```
Comment (Wescoeur) [inline rust/cli/README.md]: `List`

File: rust/cli/main.rs [hunk 3]
```diff
@@ -7,7 +7,9 @@ use xenops::*;
 fn help () {
   println!("usage:
 xenops-cli {{pause|unpause|shutdown}} <integer>
-  pause/unpause or shutdown a vm if the integer is a valid domain id.")
+  pause/unpause or shutdown a vm if the integer is a valid domain id.
+xenops-cli domain-list`
+  list all domains ids.")
```
Comment (Wescoeur) [inline rust/cli/main.rs]: I don't know if we have to do real sentences here. If you use `.`, use a capital letter too. :slightly_smiling_face:

File: rust/daemon/README.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
+HTTP server listening on port 3030 for jsonrpc request for xenops-ng.
+
+# Compile & run daemon
+
+from within the `rust` foder:
```
Comment (Wescoeur) [inline rust/daemon/README.md]: `From`

File: rust/daemon/README.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+# Examples
+
+## List domains
+
+```
+> curl -X POST -H "Content-Type: application/json" -d '{"jsonrpc": "2.0", "method": "host.domain-list", "id":123 }' <server_ip>:3030
+{"jsonrpc":"2.0","result":[0,10],"id":123}
+```
+> the result is the list of existing domain ids
```
Comment (Wescoeur) [inline rust/daemon/README.md]: `The` + `.`

File: rust/daemon/README.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+```
+> curl -X POST -H "Content-Type: application/json" -d '{"jsonrpc": "2.0", "method": "vm.pause", "params": { "dom_id": 5 }, "id": 1}' <server_ip>:3030
+{"jsonrpc":"2.0","result":"success","id":1}
+```
+
+```
+> curl -X POST -H "Content-Type: application/json" -d '{"jsonrpc": "2.0", "method": "vm.pause", "params": { "dom_id": 12 }, "id": 1}' <server_ip>:3030
+{"jsonrpc":"2.0","error":{"code":0,"message":"-3: No such process (os error 3) ()"},"id":1}
+```
+> Here the dom_id is invalid
```
Comment (Wescoeur) [inline rust/daemon/README.md]: `.`

File: rust/daemon/README.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+```
+> curl -X POST -H "Content-Type: application/json" -d '{"jsonrpc": "2.0", "method": "vm.unpause", "params": { "dom_id": 5 }, "id": 1}' <server_ip>:3030
+{"jsonrpc":"2.0","result":"success","id":1}
+```
+
+```
+> curl -X POST -H "Content-Type: application/json" -d '{"jsonrpc": "2.0", "method": "vm.unpause", "params": { "dom_id": 12 }, "id": 1}' <server_ip>:3030
+{"jsonrpc":"2.0","error":{"code":0,"message":"-3: No such process (os error 3) ()"},"id":1}
+```
+> Here the dom_id is invalid
```
Comment (Wescoeur) [inline rust/daemon/README.md]: Same.

File: rust/daemon/main.rs [hunk 8]
```diff
@@ -82,4 +83,13 @@ fn main () {
       Err(e) => Err(make_error(&e.to_string()))
     }
   } } );
+
+  let server = ServerBuilder::new(io)
+    .threads(2)
+    .rest_api(RestApi::Unsecure)
+    .cors(DomainsValidation::AllowOnly(vec![AccessControlAllowOrigin::Any]))
+    .start_http(&"0.0.0.0:3030".parse().unwrap()) // Any ip
```
Comment (Wescoeur) [inline rust/daemon/main.rs]: Same idea here, what's the convention concerning comments?

File: rust/cli/README.md (hunk 1 above)
Comment (benjamreis) [reply rust/cli/README.md]:
```suggestion
- `xenops-cli {{pause|unpause|shutdown}} <integer>`: pause/unpause/shutdown a domain, the integer arg must be a valid domain id.
```

File: rust/cli/main.rs (hunk 3 above)
Comment (benjamreis) [reply rust/cli/main.rs]:
```suggestion
  List all domains ids.")
```

File: rust/daemon/README.md (hunk 4 above)
Comment (benjamreis) [reply rust/daemon/README.md]:
```suggestion
From within the `rust` foder:
```

File: rust/daemon/README.md (hunk 5 above)
Comment (benjamreis) [reply rust/daemon/README.md]:
```suggestion
> The result is the list of existing domain ids.
```

File: rust/daemon/README.md (hunk 6 above)
Comment (benjamreis) [reply rust/daemon/README.md]:
```suggestion
> Here the dom_id is invalid.
```

File: rust/daemon/README.md (hunk 7 above)
Comment (benjamreis) [reply rust/daemon/README.md]:
```suggestion
> Here the dom_id is invalid.
```

File: rust/README.md [hunk 9]
```diff
@@ -4,6 +4,17 @@ Basic POC based on [Wendel's rust binding of libxc](https://github.com/Wenzel/xe
 
 Install `cargo` & `rustc`
 
+## Specific case: XCP-ng
```
Comment (olivierlambert) [inline rust/README.md]: I would say directly "On XCP-ng".

File: rust/daemon/main.rs (hunk 8 above)
Comment (benjamreis) [reply rust/daemon/main.rs]: you'd prefer on top?

File: rust/README.md (hunk 9 above)
Comment (benjamreis) [reply rust/README.md]:
```suggestion
## On XCP-ng
```

File: rust/daemon/main.rs (hunk 8 above)
Comment (Wescoeur) [reply rust/daemon/main.rs]: No no, capital letter + dot, or not. ^^

Comment (benjamreis) [reply rust/daemon/main.rs]:
```suggestion
    .start_http(&"0.0.0.0:3030".parse().unwrap()) // Any ip.
```

File: rust/daemon/main.rs [hunk 10]
```diff
@@ -88,7 +88,7 @@ fn main () {
     .threads(2)
     .rest_api(RestApi::Unsecure)
     .cors(DomainsValidation::AllowOnly(vec![AccessControlAllowOrigin::Any]))
-    .start_http(&"0.0.0.0:3030".parse().unwrap()) // Any ip
+    .start_http(&"0.0.0.0:3030".parse().unwrap()) // Any ip.
```
Comment (Wescoeur) [inline rust/daemon/main.rs:91]: :fireworks:

**Review by Wescoeur — APPROVED** [review]

**Pull request merged**